src/Plugin/search_api/processor/TypedRelationFiltered.php

```diff
@@ -0,0 +1,121 @@
+<?php
+
+namespace Drupal\controlled_access_terms\Plugin\search_api\processor;
+
+use Drupal\controlled_access_terms\Plugin\search_api\processor\Property\TypedRelationFilteredProperty;
+use Drupal\search_api\Datasource\DatasourceInterface;
+use Drupal\search_api\Item\ItemInterface;
+use Drupal\search_api\Processor\ProcessorPluginBase;
+
+/**
+ * Adds filterable fields for each Typed Relation field.
+ *
+ * @SearchApiProcessor(
+ *   id = "typed_relation_filtered",
+ *   label = @Translation("Typed Relation, filtered by type"),
+ *   description = @Translation("Filter Typed Relation fields by type"),
+ *   stages = {
+ *     "add_properties" = 0,
+ *   },
+ *   locked = true,
+ *   hidden = false,
+ * )
+ */
+class TypedRelationFiltered extends ProcessorPluginBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getPropertyDefinitions(DatasourceInterface $datasource = NULL): array {
+    $properties = [];
+
+    if (!$datasource || !$datasource->getEntityTypeId()) {
+      return $properties;
+    }
+
+    $entity_type = $datasource->getEntityTypeId();
+    // Get all configured typed relation fields.
+    $fields = \Drupal::entityTypeManager()->getStorage('field_config')->loadByProperties([
+      'entity_type' => $entity_type,
+      'field_type' => 'typed_relation',
```

Comment:
Would it be possible to store the field name in here so the logic when it comes around to be used is simple/less expensive?